### ISSUE TITLE
[DocDB] Transaction status tablets seem to always get placed in the cluster's default placement info even when transaction takes place entirely in a tablespace

### ISSUE THREAD

**Issue by jteplitz** [yugabyte/yugabyte-db#16021]
### Description

**Overview**

When deploying a multi-region cluster with 5 nodes in region A and 5 nodes in region B, we've set the default placement info to the 5 nodes in region A. We then created a tablespace with placement info in region B and created tables and indices on that tablespace. Single row transactions in region B are fast and local to region B. However, we've observed that multi-row transactions, such as inserts into a table with an index, are always using a transaction status tablet in region A. Since the latency between these two regions is large (100s of m.s.) this leads to very bad write latencies.

**Steps to reproduce**
1. Deploy 10 nodes across two regions.
2. Run `yb-admin modify_placement_info` to set the default placement info to the 5 nodes in region A.
3. Create a ysql tablespace, `my-tablespace`, with placement info in region B
4. Execute the following sql:
```sql
CREATE TABLE test (username tsvector) TABLESPACE my-tablespace;
CREATE INDEX on test USING gin(username) TABLESPACE my-tablespace;
```
5. Connect to a node in region B and insert rows into `test` repeatedly.

**Results**
Every 1 second, the node that that you're connected to in region B will print the following warning log message:
```No local transaction status tablet found```

Monitoring network traffic on that node via tcdump, will show periodic traffic to nodes in region A alongside on each insert query.
You can add a breakpoint on `PickStatusTabletId` and print `tablets` to see the status tablets that are available for the transaction. Running `yb-admin list_tablet_servers` on these tablets shows that they are all in region A.

**Comment by ddorian** [comment]
@jteplitz 

Please add the query used to create the tablespace?

**Comment by ddorian** [comment]
On https://docs.yugabyte.com/preview/admin/yb-admin/#set-preferred-zones:

> Tablespaces don't inherit cluster-level placement information, leader preference, or read replica configurations.

Did you mean the preferred_zones for this tablespace too? Or was it for other tables?

Please also specify the full command you used for `yb-admin modify_placement_info` and `yb-admin set_preferred_zones`.

**Comment by deeps1991** [comment]
@jteplitz Thanks a lot for such detailed information presented above. I have just one other question. Would you be able to look at the YB-Master UI page for your cluster and go to the "Tables" page? (For example on my local cluster, the page is in `http://127.0.0.1:7000/tables`).
That page lists all the tables in the cluster. Whenever a tablespace is created, there should be an associated `transaction status table` created for the tablespace with the same placement policy.

On the Tables UI page, you should see a table under System Tables called `transactions_<uuid>` that corresponds to the tablespace test, and if you click it, you should see that its `Replication Info` matches that of the tablespace test. Could you let me know if you see such a transactions table and if its placement info looks correct?

@es1024 One interesting point in the above use case is that the placement information of B is not a subset of the default placement policy + default preferred leader config. Not sure if that is the cause of the issue, but might be good to repro such a case and see.

**Comment by jteplitz** [comment]
@deeps1991 We have quite a few tablespaces in this deployment. Is there a way to get the tablespace UUID (I only see the OID on the table's page)?

I do see a `transactions_<uuid>` table with the same placement info as this tablespace though, so I'm guessing that's the one.

**Comment by es1024** [comment]
@jteplitz The UUID in the transaction table name is randomly generated, and unfortunately I don't think we maintain a mapping from tablespace to transaction table anywhere (the mapping is stored in the reverse direction only -- each transaction table has a tablespace id). But if you don't have multiple tablespaces with the same placement info, matching placement info should be enough.

Can you verify that each of the tservers in region B leads at least one tablet of the `transactions_<uuid>` table corresponding to the tablespace? 

If that is indeed the case, then the transaction you're running is being probably started as a global transaction rather than a region-local transaction, which means that a transaction status tablet from the `system.transactions` table will be used. But since your cluster config is entirely in region A, and `system.transactions` always follows the cluster config, this means that nodes in region B will never lead a tablet from this table, hence the `No local transaction status tablet found` warning.

**Comment by bmatican** [comment]
@es1024 should we cut a task for this then?
> I don't think we maintain a mapping from tablespace to transaction table anywhere

**Comment by jteplitz** [comment]
@es1024 Yup, each tablet server in region B is a leader for at least one tablet in this `transactions_<uuid>` table.

**Comment by jteplitz** [comment]
Out of curiosity, can you explain when a transaction should use the tablespace transaction tablet? Looking over `src/yb/client/transaction.cc`, it looks a bit like any transaction with one operation on a tablet outside of the current region will get promoted to global:

```c++
  internal::InFlightOp* FindOpWithLocalityViolation(
      internal::InFlightOpsGroupsWithMetadata* ops_info) REQUIRES(mutex_) {
    if (metadata_.locality != TransactionLocality::LOCAL) {
      return nullptr;
    }
    for (auto& group : ops_info->groups) {
      auto& first_op = *group.begin;
      auto tablet = first_op.tablet;
      if (!tablet->IsLocalRegion()) {
        return &first_op;
      }
    }
    return nullptr;
  }
```

**Comment by es1024** [comment]
> Out of curiosity, can you explain when a transaction should use the tablespace transaction tablet?
The only time it is used is indeed when all operations of the transaction are within the current region. If the transaction starts with non region local operations, it will be started as a global transaction using a tablet of the `system.transactions` table, and if the transaction only later on runs into a non region local operation, it will move to using a tablet from the `system.transactions` table.

I just realized that in your case, you specified region A and region B, but those don't actually correspond to single regions in your setup. So the transaction tables created with your tablespaces won't actually be used (#11268). Your observation that single row transactions are fast is due to [optimizations for single row transactions](https://docs.yugabyte.com/preview/architecture/transactions/single-row-transactions/).

**Comment by es1024** [comment]
> @es1024 should we cut a task for this then?
> 
> > I don't think we maintain a mapping from tablespace to transaction table anywhere
@bmatican It would be nice to have, but I'm not sure we have a good place to put it at the moment, since tablespaces are handled almost entirely on the PG side (e.g. pretty much the only thing we display for tablespaces in the master UI is the tablespace ID of each table).

**Comment by bmatican** [comment]
@es1024 Fair point, we can discuss that, eg if it should be some internal pg function, or some new master U, since the CatalogManager also refreshes the PG tablespaces, for load balancing and such, so it would be able to build the reverse mapping pretty easily.

Anyway, probably lower pri, but figured to keep track of the gap as an issue.

**Comment by jteplitz** [comment]
@es1024 Thanks for the explanation about region local vs. global transactions! I understand what you're saying now and the placement policy is really a red herring. The real issue is that all of our transactions will be global like you say and the global txn table is either far away (with the default placement policy) or potentially sharded across tablets that have very high communication latencies (without the default placement policy).

For my own understanding of https://github.com/yugabyte/yugabyte-db/issues/11268, is the main reason xregion transactions get promoted to global due to latency concerns with the region local transaction table? Or are their correctness concerns with using the regional transaction table for such a transaction?

**Comment by es1024** [comment]
@jteplitz The reason we don't use a region local transaction table when there are writes outside of the region is to prevent an outage in that region from causing an outage outside of the region: if a transaction writing to data in both region A and B uses a transaction table replicated only within region A, and region A becomes unavailable after the transaction commits, but before its intents have resolved, then any later transaction attempting to read the data written by the transaction to region B will be stuck, since it'd have to check if the transaction had committed or not, and this information is only available in the transaction table in region A (which is unavailable).

**Comment by jteplitz** [comment]
Gotcha, that makes sense. Thanks for the detailed explanation!
